### PR TITLE
Support MTP loss mask_type v1 and multi mtp_config

### PR DESCRIPTION
**Pull request by x54-729**
mtp config example:
```python
model_cfg.text_config.mtp_config = [
    MTPConfig(name="normal", num_layers=TEXT_MTP_LAYERS, share_weights=TEXT_MTP_LAYERS>1, loss_scaling_factor=TEXT_MTP_FACTOR),
    MTPConfig(name="sci", num_layers=NUM_MTP_LAYERS, share_weights=NUM_MTP_LAYERS>1, loss_scaling_factor=NUM_MTP_FACTOR, loss_cfg=SciMTPLossConfig(mask_type="v1"))
]
```
Refer to https://github.com/HAOCHENYE/xtuner/pull/2

### CODE REVIEW

File: xtuner/v1/engine/train_engine.py [hunk 1]
```diff
@@ -569,6 +569,9 @@ def _get_total_loss(self, model_outputs: ModelOutputs) -> torch.Tensor:
         loss = torch.tensor(0.0, device=DEVICE)
         for key in model_outputs.model_fields:
             value = getattr(model_outputs, key)
-            if "loss" in key and isinstance(value, torch.Tensor):
```
Comment (HAOCHENYE) [inline xtuner/v1/engine/train_engine.py:572]: The loss part should be autonomous to the model, rather than being hardcoded here.

File: xtuner/v1/loss/mtp_loss.py [hunk 2]
```diff
@@ -167,6 +168,12 @@ def forward(
             head_weight = head_weight.detach()
             head_bias = head_bias.detach() if head_bias is not None else None
         # Dispatch to eager_mode/chunk_mode via base class, which calls loss_fn per chunk
+
+        mask_type = self.loss_cfg.mask_type
+        if mask_type == "v1":
+            self.process_loss_weight_v1()
+        elif mask_type is not None:
+            raise NotImplementedError(f"Unknown MTP Loss Mask Type: {mask_type}")
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/mtp_loss.py]: The calculation logic of loss should not be hard-coded here; please implement a new loss_context.

File: xtuner/v1/loss/mtp_loss.py [hunk 3]
```diff
@@ -214,3 +221,34 @@ def _kl_loss_fn(
         )
 
         return kl_loss, (None, {})
+
+    def process_loss_weight_v1(self):
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/mtp_loss.py:287]: The loss calculation logic here is no longer a generic mtp head. Please implement a new loss_context.

File: xtuner/v1/model/moe/moe.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 import os
 import types
 from pathlib import Path
-from typing import TYPE_CHECKING, Annotated, Literal, Self, Sequence, TypedDict, cast
+from typing import TYPE_CHECKING, Annotated, List, Literal, Self, Sequence, TypedDict, cast
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: Use python builtin `list`

File: xtuner/v1/engine/train_engine.py [hunk 5]
```diff
@@ -569,6 +569,9 @@ def _get_total_loss(self, model_outputs: ModelOutputs) -> torch.Tensor:
...
+            if key == "mtp_loss" and isinstance(value, dict):
+                for mtp_loss_name, mtp_loss in value.items():
+                    loss += mtp_loss
+            elif "loss" in key and isinstance(value, torch.Tensor):
```
Comment (HAOCHENYE) [inline xtuner/v1/engine/train_engine.py]:
```suggestion
            elif "loss" in key:
                loss_values = list(value.values()) if isinstance(value, dict) else [value]
                loss_values = [i for i in loss_values if isinstance(i, torch.Tensor)]
                for value in loss_values:
                    loss += value
```

File: xtuner/v1/model/moe/moe.py [hunk 6]
```diff
@@ -645,6 +687,69 @@ def _micro_batch_forward(
...
+        self,
+        mtp_config: MTPConfig,
+        output,
+        layer_hidden_states,
+        position_embeddings,
+        balancing_ctx,
+        z_ctx,
+        mtp_seq_ctx,
+        mtp_loss_ctx_dict,
+        keep_router: bool,
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: missing typehint here

File: xtuner/v1/loss/mtp_loss.py [hunk 7]
```diff
@@ -53,24 +53,32 @@ class MTPLossConfig(CELossConfig):
...
+    mtp_depth: int = 1
     detach_mtp_lm_head_weight: bool = False
 
+    def bind_mtp_depth(self, idx: int) -> None:
+        """Bind MTP depth to the given index.
+
+        Args:
+            idx (int): 1-indexed MTP layer depth to bind.
+        """
+        self.mtp_depth = idx
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/mtp_loss.py]: Depth and layer count are two different things. The user configures the number of layers in the config, whereas the layer index — which layer this actually is — is known at construction time. I'd suggest moving the `bind_mtp_depth` logic into `LossContext`, so that a single `MTPConfig` can produce `LossContext` instances for the different layers.

File: xtuner/v1/model/moe/moe.py [hunk 8]
```diff
@@ -751,59 +856,31 @@ def _forward(
...
-                    .contiguous()
-                    .float(),
-                    selected_router_logits=mtp_router_results.index_select(0, mtp_nonpad_indices).contiguous().float(),
-                    hidden_states=mtp_hidden_states,
+            for mtp_config in self.config.mtp_config:
+                self._mtp_forward(
+                    mtp_config=mtp_config,
+                    output=output,
+                    layer_hidden_states=layer_hidden_states,
+                    position_embeddings=position_embeddings,
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: Let's keep this part of the code unchanged for now — we shouldn't be doing code cleanup/refactoring in this PR.

File: xtuner/v1/model/moe/moe.py [hunk 9]
```diff
@@ -1084,28 +1181,38 @@ def fully_shard(
...
-                    module=mtp_layer,
-                )
-                if mtp_idx == 0:
-                    layer_next.set_modules_to_forward_prefetch([mtp_layer])  # type: ignore
-
-            if self.config.mtp_config is not None and self.config.mtp_config.num_layers > 0:
+            global_mtp_idx = 0  # Track global MTP layer index across all mtp_configs
+            for mtp_name in self.mtp_block.keys():
+                mtp_block = self.mtp_block[mtp_name]
+                mtp_config = next((cfg for cfg in self.config.mtp_config if cfg.name == mtp_name), None)  # type: ignore
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: Just pass `mtp_config` directly into MTPBlock — no need for all this indirection.

File: xtuner/v1/model/moe/moe.py [hunk 10]
```diff
@@ -1084,28 +1181,38 @@ def fully_shard(
...
+                for local_mtp_idx, mtp_layer in enumerate(mtp_block.layers):
+                    if self._should_recompute(None, mtp_idx=global_mtp_idx) or (
+                        mtp_config is not None and mtp_config.share_weights
+                    ):  # share mtp head must recompute
+                        mtp_layer = checkpoint_wrapper(mtp_layer, checkpoint_impl=CheckpointImpl.REENTRANT)
+                    mtp_block.layers[local_mtp_idx] = mtp_layer
+
+                    reshard_after_forward = local_mtp_idx != len(mtp_block.layers) - 1
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: `reshard_after_forward` should only be set to `True` for the last layer of the last mtp block

File: xtuner/v1/model/moe/moe.py [hunk 11]
```diff
@@ -1084,28 +1181,38 @@ def fully_shard(
...
+            global_mtp_idx = 0  # Track global MTP layer index across all mtp_configs
+            for mtp_name in self.mtp_block.keys():
+                mtp_block = self.mtp_block[mtp_name]
+                mtp_config = next((cfg for cfg in self.config.mtp_config if cfg.name == mtp_name), None)  # type: ignore
+                for local_mtp_idx, mtp_layer in enumerate(mtp_block.layers):
+                    if self._should_recompute(None, mtp_idx=global_mtp_idx) or (
+                        mtp_config is not None and mtp_config.share_weights
+                    ):  # share mtp head must recompute
+                        mtp_layer = checkpoint_wrapper(mtp_layer, checkpoint_impl=CheckpointImpl.REENTRANT)
+                    mtp_block.layers[local_mtp_idx] = mtp_layer
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]:
```suggestion
                        mtp_layer = checkpoint_wrapper(mtp_layer, checkpoint_impl=CheckpointImpl.REENTRANT)
                        mtp_block.layers[local_mtp_idx] = mtp_layer
```

File: xtuner/v1/model/moe/moe.py [hunk 12]
```diff
@@ -202,7 +209,7 @@ def __init__(self, config: MoEConfig):
         self.layers = self.build_layers(config)
         self.rotary_emb = self.build_rotary_embedding(config)
         self.embed_tokens = self.build_embeddings(config)
-        self.mtp_block = self.build_mtp_block(config) if config.mtp_config is not None else None
+        self.mtp_block = self.build_mtp_block_dict(config) if config.mtp_config is not None else None
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: Could we switch this to a `ModuleList` too? That way it'd be symmetric with the config.
